File: go/cryptoutil/x25519.go
package cryptoutil

import (
	"crypto/ecdh"
	"crypto/rand"
	"github.com/jc-lab/jscp/go/payloadpb"
	"github.com/pkg/errors"
)

var x25519Algorithm X25519Algorithm

type X25519Algorithm struct {
}

func (a *X25519Algorithm) GetKeyFormat() payloadpb.KeyFormat {
	return payloadpb.KeyFormat_KeyFormatX25519
}

func (a *X25519Algorithm) GetType() payloadpb.DHAlgorithm {
	return payloadpb.DHAlgorithm_DHX25519
}

func (a *X25519Algorithm) Generate() (*DHKeyPair, error) {
	curve := ecdh.X25519()
	privateKey, err := curve.GenerateKey(rand.Reader)
	if err != nil {
		return nil, err
	}
	keyPair := &DHKeyPair{
		Private: &X25519PrivateKey{
			Key: privateKey,
		},
		Public: &X25519PublicKey{
			Key: privateKey.PublicKey(),
		},
	}
	return keyPair, nil
}

func (a *X25519Algorithm) UnmarshalPublicKey(input []byte) (PublicKey, error) {
	return a.UnmarshalDHPublicKey(input)
}

func (a *X25519Algorithm) UnmarshalDHPublicKey(input []byte) (DHPublicKey, error) {
	curve := ecdh.X25519()
	publicKey, err := curve.NewPublicKey(input)
	if err != nil {
		return nil, err
	}
	return &X25519PublicKey{
		Key: publicKey,
	}, nil
}

type X25519PrivateKey struct {
	Key *ecdh.PrivateKey
}

func (k *X25519PrivateKey) IsDHKey() bool {
	return true
}

func (k *X25519PrivateKey) IsSignatureKey() bool {
	return false
}

func (k *X25519PrivateKey) GetDHPublic() DHPublicKey {
	return &X25519PublicKey{
		Key: k.Key.PublicKey(),
	}
}

func (k *X25519PrivateKey) GetPublic() PublicKey {
	return k.GetDHPublic()
}

func (k *X25519PrivateKey) Algorithm() PublicAlgorithm {
	return &x25519Algorithm
}

func (k *X25519PrivateKey) GetDHAlgorithm() DHAlgorithm {
	return &x25519Algorithm
}

func (k *X25519PrivateKey) GetDHAlgorithmProto() payloadpb.DHAlgorithm {
	return payloadpb.DHAlgorithm_DHX25519
}

func (k *X25519PrivateKey) DH(peerKey DHPublicKey) ([]byte, error) {
	publicKey, ok := peerKey.(*X25519PublicKey)
	if !ok {
		return nil, errors.New("peerKey is not x25519")
	}
	return k.Key.ECDH(publicKey.Key)
}

type X25519PublicKey struct {
	Key *ecdh.PublicKey
}

func (k *X25519PublicKey) IsDHKey() bool {
	return true
}

func (k *X25519PublicKey) IsSignatureKey() bool {
	return false
}

func (k *X25519PublicKey) Algorithm() PublicAlgorithm {
	return &x25519Algorithm
}

func (k *X25519PublicKey) GetDHAlgorithm() DHAlgorithm {
	return &x25519Algorithm
}

func (k *X25519PublicKey) GetDHAlgorithmProto() payloadpb.DHAlgorithm {
	return payloadpb.DHAlgorithm_DHX25519
}

func (k *X25519PublicKey) Marshal() []byte {
	return k.Key.Bytes()
}

func (k *X25519PublicKey) MarshalToProto() (*payloadpb.PublicKey, error) {
	return &payloadpb.PublicKey{
		Format: payloadpb.KeyFormat_KeyFormatX25519,
		Data:   k.Marshal(),
	}, nil
}

File: node/src/crypto/x25519.ts
import * as x25519 from '@stablelib/x25519';
import {
    PublicAlgorithm,
    DHAlgorithm,
    PublicKey,
    DHPublicKey,
    DHKeyPair,
    DHPrivateKey
} from './types';
import * as proto from '../proto';

export class X25519PrivateKey implements DHPrivateKey {
    constructor(
        private readonly key: Uint8Array,
    ) {
        if (key.length != x25519.SECRET_KEY_LENGTH) {
            throw new Error(`invalid key (size=${key.length})`);
        }
    }

    // Key

    isDHKey(): boolean {
        return true;
    }

    isSignatureKey(): boolean {
        return false;
    }

    algorithm(): PublicAlgorithm {
        return x25519Algorithm;
    }

    // DHPrivateKey

    getPublic(): PublicKey {
        return this.toPublic();
    }


    getDHAlgorithm(): DHAlgorithm {
        return x25519Algorithm;
    }

    getDHAlgorithmProto(): proto.DHAlgorithm {
        return proto.DHAlgorithm.DHX25519;
    }

    getDHPublic(): DHPublicKey {
        return this.toPublic();
    }

    async dh(peerKey: DHPublicKey): Promise<Uint8Array> {
        if (!(peerKey instanceof X25519PublicKey)) {
            throw new Error('peerKey is not X25519PublicKey');
        }
        return x25519.sharedKey(this.key, peerKey.key);
    }

    toPublic(): X25519PublicKey {
        return new X25519PublicKey(
            x25519.generateKeyPairFromSeed(this.key).publicKey
        )
    }
}

export class X25519PublicKey implements DHPublicKey {
    constructor(public readonly key: Uint8Array) {
        if (key.length != x25519.PUBLIC_KEY_LENGTH) {
            throw new Error(`invalid key (size=${key.length})`);
        }
    }

    // Key

    isDHKey(): boolean {
        return true;
    }

    isSignatureKey(): boolean {
        return false;
    }

    algorithm(): PublicAlgorithm {
        return x25519Algorithm;
    }

    // PublicKey

    marshalToProto(): proto.PublicKey {
        return {
            format: proto.KeyFormat.KeyFormatX25519,
            data: this.key,
        };
    }

    // DHPublicKey

    getDHAlgorithm(): DHAlgorithm {
        return x25519Algorithm;
    }

    getDHAlgorithmProto(): proto.DHAlgorithm {
        return proto.DHAlgorithm.DHX25519
    }

    marshal(): Uint8Array {
        return this.key;
    }
}

export class X25519Algorithm implements DHAlgorithm {
    // PublicAlgorithm
    getKeyFormat(): proto.KeyFormat {
        return proto.KeyFormat.KeyFormatX25519;
    }

    async unmarshalPublicKey(input: Uint8Array): Promise<PublicKey> {
        return new X25519PublicKey(input);
    }

    // DHAlgorithm

    getType(): proto.DHAlgorithm {
        return proto.DHAlgorithm.DHX25519;
    }

    async generate(): Promise<DHKeyPair> {
        const keyPair = x25519.generateKeyPair();

        return {
            private: new X25519PrivateKey(keyPair.secretKey),
            public: new X25519PublicKey(keyPair.publicKey)
        };
    }

    async unmarshalDHPublicKey(input: Uint8Array): Promise<DHPublicKey> {
        return new X25519PublicKey(input);
    }
}

export const x25519Algorithm = new X25519Algorithm();
